File: java/dp.java
public class dp {
    

    public int wizard_magic(int[] m, int[] t, int H, int n) {
        int[] dp = new int[H + 1];  // dp array to store results
        
        // bottom up loop
        for (int hours = 0; hours <= H; hours++) {
            // try each potion, save the max value
            int max_subproblem = 0;
            for (int potion = 0; potion < n; potion++) {
                if (hours - t[potion] >= 0) {  // ensure that we have time
                    max_subproblem = Math.max(max_subproblem, 
                                             m[potion] + dp[hours - t[potion]]);
                }
            }
            dp[hours] = max_subproblem;
        }
        return dp[H];
    }


    public interface MagicSolver {
        int solve(int[] m, int[] t, int H, int n);
    }

    
    public static void main(String[] args) {
        dp instance = new dp(); 
        small_tests tester = new small_tests();
        System.out.println("running small tests");
        tester.small_test((m, t, H, n) -> instance.wizard_magic(m, t, H, n));
        big_tests btester = new big_tests();
        System.out.println("\nrunning big tests");
        btester.big_test((m, t, H, n) -> instance.wizard_magic(m, t, H, n));
    }

    
}

File: python/dp.py


def wizard_magic(m, t, H, n):
    
    dp = [0]*(H+1)
    
    # bottom up loop
    for hours in range(H+1):
        # try each potion, save the max value
        max_subproblem = 0
        for potion in range(n):
            if hours - t[potion] >= 0:  # ensure that we have time
                max_subproblem = max(max_subproblem, 
                                     m[potion] + dp[hours - t[potion]])
        dp[hours] = max_subproblem
    return dp[H]
# =============================================================================
# ============================================================================= 
if __name__ == "__main__":
    import time
    from small_tests import small_test
    print('\nrunning first tests')
    time.sleep(0.5)
    small_test(wizard_magic) 
    
    from big_tests import big_test
    time.sleep(0.25)
    print('\nrunning bigger tests')
    time.sleep(0.25)
    time.sleep(0.5)
    big_test(wizard_magic) 


   
   